pkg/controllers/pushsecret/pushsecret_controller_test.go

```diff
@@ -1012,3 +1020,325 @@ var _ = Describe("ExternalSecret controller", func() {
+				secretValue := secret.Data[defaultKey]
+				providerValue, ok := fakeProvider.SetSecretArgs[ps.Spec.Data[0].Match.RemoteRef.RemoteKey]
+				if !ok {
+					return false
+				}
+				got := providerValue.Value
+				return bytes.Equal(got, secretValue)
+			}, time.Second*10, time.Second).Should(BeTrue())
+			return true
+		}
+	}
+
+	noManagedStoresFails := func(tc *testCase) {
+		fakeProvider.SetSecretFn = func() error {
+			return nil
+		}
+
+		tc.pushsecret.Spec.SecretStoreRefs = []v1alpha1.PushSecretStoreRef{
+			{
+				Name: UnmanagedPushSecretStore1,
+				Kind: "SecretStore",
+			},
+			{
+				Name: UnmanagedPushSecretStore2,
+				Kind: "SecretStore",
+			},
+		}
+
+		tc.assert = func(ps *v1alpha1.PushSecret, secret *v1.Secret) bool {
+			expected := v1alpha1.PushSecretStatusCondition{
+				Type:    v1alpha1.PushSecretReady,
+				Status:  v1.ConditionFalse,
+				Reason:  v1alpha1.ReasonErrored,
+				Message: "PushSecret \"test-ps\" has no managed stores to push to",
+			}
+			return checkCondition(ps.Status, expected)
+		}
+	}
+
+	warnUnmanagedStoresAndSyncManagedStores := func(tc *testCase) {
+		fakeProvider.SetSecretFn = func() error {
+			return nil
+		}
+
+		tc.pushsecret.Spec.SecretStoreRefs = []v1alpha1.PushSecretStoreRef{
+			{
+				Name: ManagedPushSecretStore1,
+				Kind: "SecretStore",
+			},
+			{
+				Name: ManagedPushSecretStore2,
+				Kind: "SecretStore",
+			},
+			{
+				Name: UnmanagedPushSecretStore1,
+				Kind: "SecretStore",
+			},
+			{
+				Name: UnmanagedPushSecretStore2,
+				Kind: "SecretStore",
+			},
+		}
+
+		tc.assert = func(ps *v1alpha1.PushSecret, secret *v1.Secret) bool {
+			Eventually(func() bool {
+				By("checking if Provider value got updated")
+				secretValue := secret.Data[defaultKey]
+				providerValue, ok := fakeProvider.SetSecretArgs[ps.Spec.Data[0].Match.RemoteRef.RemoteKey]
+				if !ok {
+					return false
+				}
+				got := providerValue.Value
+				return bytes.Equal(got, secretValue)
+			}, time.Second*10, time.Second).Should(BeTrue())
+			return true
+		}
+	}
+
+	DescribeTable("When reconciling a PushSecret with multiple secret stores",
+		func(tweaks ...testTweaks) {
+			tc := makeDefaultTestcase()
+			for _, tweak := range tweaks {
+				tweak(tc)
+			}
+			ctx := context.Background()
+			By("creating secret stores, a secret and a pushsecret")
+			if tc.managedStore1 != nil {
+				Expect(k8sClient.Create(ctx, tc.managedStore1)).To(Succeed())
+			}
+			if tc.managedStore2 != nil {
+				Expect(k8sClient.Create(ctx, tc.managedStore2)).To(Succeed())
+			}
+			if tc.unmanagedStore1 != nil {
+				Expect(k8sClient.Create(ctx, tc.unmanagedStore1)).To(Succeed())
+			}
+			if tc.unmanagedStore2 != nil {
+				Expect(k8sClient.Create(ctx, tc.unmanagedStore2)).To(Succeed())
+			}
+			if tc.secret != nil {
+				Expect(k8sClient.Create(ctx, tc.secret)).To(Succeed())
+			}
+			if tc.pushsecret != nil {
+				Expect(k8sClient.Create(ctx, tc.pushsecret)).Should(Succeed())
+			}
+			time.Sleep(2 * time.Second)
```

Comment:
Please don't include sleeps in unit tests. There need to be a way to use eventually or something. This lengthens the unit test runtime by a hard two seconds.

Reply:
This was copied from https://github.com/external-secrets/external-secrets/blob/main/pkg/controllers/pushsecret/pushsecret_controller_test.go#L980

I have tried removing them but this is what we get from `make test`:
```
Ran 23 of 23 Specs in 98.699 seconds
SUCCESS! -- 23 Passed | 0 Failed | 0 Pending | 0 Skipped
    testing.go:1398: race detected during execution of test
--- FAIL: TestAPIs (98.70s)
FAIL
coverage: 22.0% of statements
FAIL    github.com/external-secrets/external-secrets/pkg/controllers/pushsecret 98.942s
=== RUN   TestManagerGet
=== RUN   TestManagerGet/creates_a_new_client_from_storeRef_and_stores_it
=== RUN   TestManagerGet/creates_a_new_client_using_both_storeRef_and_sourceRef
=== RUN   TestManagerGet/retrieve_cached_client_when_store_matches
=== RUN   TestManagerGet/create_new_client_when_store_doesn't_match
--- PASS: TestManagerGet (0.01s)
    --- PASS: TestManagerGet/creates_a_new_client_from_storeRef_and_stores_it (0.00s)
    --- PASS: TestManagerGet/creates_a_new_client_using_both_storeRef_and_sourceRef (0.00s)
    --- PASS: TestManagerGet/retrieve_cached_client_when_store_matches (0.00s)
    --- PASS: TestManagerGet/create_new_client_when_store_doesn't_match (0.00s)
=== RUN   TestAPIs
Running Suite: Controller Suite - /home/bude/repos/GitHub/external-secrets/pkg/controllers/secretstore
```

I can leave a comment saying it's to prevent race conditions during tests?

Reply:
Ah. I see. Leave it then. 🤣